### PR TITLE
new rule: iframe with negative tabindex has no interactive elements

### PR DESCRIPTION
**Pull request by WilcoFiers**
This rule checks that `iframe` elements with a negative `tabindex` attribute value contain no interactive elements.

Need for Final Call: 2 weeks
## How to Review And Approve

- Go to the “Files changed” tab
- Here you will have the option to leave comments on different lines.
- Once the review is completed, find the “Review changes” button in the top right, select “Approve” (if you are really confident in the rule) or "Request changes" and click “Submit review”.
- Make sure to also review the proposed Final Call period. In case of disagreement, the longer period wins.

### CODE REVIEW

**Review by ajanec01 — APPROVED** [review]

File: _rules/iframe-not-focusable-has-no-interactive-content-akn7bn.md
Comment (kasperisager) [inline _rules/iframe-not-focusable-has-no-interactive-content-akn7bn.md:33]: "includes elements" needs to be qualified to ensure unambiguity. I'm guessing we're looking for "has descendants in the flat tree"?

**Review by kasperisager — DISMISSED** [review]

**Review by carlosapaduarte — CHANGES_REQUESTED** [review]
Besides @kasperisager's comment I believe an inapplicable example without a tabindex attribute could also be useful.

File: _rules/iframe-not-focusable-has-no-interactive-content-akn7bn.md
Comment (carlosapaduarte) [inline _rules/iframe-not-focusable-has-no-interactive-content-akn7bn.md:33]:
```suggestion
None of the test target's [nested browsing context][] includes elements that are [visible][] and part of the [sequential focus navigation][]. An element is "included" in a [nested browsing context][] if its [owner document][] is the [container document][] of the [nested browsing context][].
```

**Review by carlosapaduarte — APPROVED** [review]
Just an edit suggestion to improve the clarity, but I'm approving this anyway.

**Review by kasperisager — DISMISSED** [review]

**Comment by wcreedle** [comment]
Forgive me if this is not the place to ask this example (first time dipping my toes in the water), but how would this rule handle an example like this?

Issue: there is a non conforming video type that has pertinent information to the user. As a workaround the page
1. sets negative tabindex on the video
2. provides an accessible transcript in front of the video to provide the same information.

File: _rules/iframe-not-focusable-has-no-interactive-content-akn7bn.md
Comment (Jym77) [inline _rules/iframe-not-focusable-has-no-interactive-content-akn7bn.md:33]: As far as I understand, an `iframe` only has one nested browsing context, not several (so "none of" is not needed).

Comment (Jym77) [inline _rules/iframe-not-focusable-has-no-interactive-content-akn7bn.md:33]: Linked to previous comment, I understand the expectation as not going recursively into nested `iframe`, because "owner document" only look at the parent, not the top-level ancestor.
I don't know if this is intentional or not.

**Review by Jym77 — DISMISSED** [review]

**Comment by WilcoFiers** [comment]
@wcreedle Thanks for the input. This is a little complicated, but I'll try to explain why I think this is correct. WCAG allows a page to conform, even if a success criterion is not satisfied, as long as a conforming alternative version is provided. This rule tests specifically for the success criterion; so the rule finds things that do not satisfy the success criterion, but even with content that does not satisfy the SC, a page may still conform if an alternative version is provided.

**Comment by wcreedle** [comment]
Thank you, this helps a lot.

File: _rules/iframe-not-focusable-has-no-interactive-content-akn7bn.md
Comment (WilcoFiers) [reply _rules/iframe-not-focusable-has-no-interactive-content-akn7bn.md:33]: It's not recursive, no. Because `iframe` elements are themselves focusable, an empty `iframe` inside your `iframe[tabindex="-1"]` would fail this rule unless it too had a negative tabindex. That is captured in the assumption. I think that's reasonable. Iframes need to be focusable in order to allow scrolling with keyboard.

Comment (WilcoFiers) [reply _rules/iframe-not-focusable-has-no-interactive-content-akn7bn.md:33]: Fair point, I'll fix that.

Comment (Jym77) [reply _rules/iframe-not-focusable-has-no-interactive-content-akn7bn.md]: 🤔 
```html
<iframe tabindex="-1" srcdoc="<iframe tabindex='-1' srcdoc='<a href=…>W3C</a>' />" />
```
(not filling the `href` to avoid quote protection nightmare)
Then the link is not keyboard accessible (the SC is not satisfied) but the rule would pass because it only looks at the outer `iframe` and is happy to see that none of its elements are part of sequential focus navigation.

And I don't think this is caught by the assumption since the iframe (both inner and outer) do indeed provide functionality.

This is not a blocker, though since it means the rule passes too often, which is usual.

**Review by carlosapaduarte — CHANGES_REQUESTED** [review]
Editorial

File: _rules/iframe-not-focusable-has-no-interactive-content-akn7bn.md
Comment (Jym77) [reply _rules/iframe-not-focusable-has-no-interactive-content-akn7bn.md]: Ah no, my bad. The inner iframe should be applicable and fail.
It is not completely explicit in the applicability, however…

File: _rules/iframe-not-focusable-has-no-interactive-content-akn7bn.md
Comment (ajanec01) [inline _rules/iframe-not-focusable-has-no-interactive-content-akn7bn.md:33]: Wouldn't that first line be more appropriate in the applicability section? At the moment the applicability says all `<iframe tabindex="-1">` whereas it's more about `<iframe tabindex="-1">` with widgets.

Comment (WilcoFiers) [reply _rules/iframe-not-focusable-has-no-interactive-content-akn7bn.md:33]: I'm not sure. I suppose it can be done in a bunch of different ways. We could also completely flip it around and make this applicable to any focusable element in an iframe, and then have an expectation that the iframe doesn't have a tabindex.... I think I prefer it this way around, but if you have a strong preference I don't mind changing it.

Comment (ajanec01) [reply _rules/iframe-not-focusable-has-no-interactive-content-akn7bn.md:33]: No strong preference. Valid point.

**Review by ajanec01 — APPROVED** [review]

**Review by kasperisager — APPROVED** [review]

**Review by carlosapaduarte — APPROVED** [review]

**Review by Jym77 — COMMENTED** [review]
Still good. Only commenting since we already have one SI approval.

**Comment by WilcoFiers** [comment]
Putting it back into final call, ending on January 29th.

**Pull request merged**
